Step3 (JEJE)

jung-yun opened the pull request:
Player struct를 class로 바꿔서 오브젝트의 값을 바꿀 수 있도록 했습니다. (값 타입과 레퍼런스 타입의 차이인 것 같습니다.)
```
for player in playersArray {
                if let drawnCard = cardDeck.removeOne() {
                    player.receive(card: drawnCard)
                } else {
                    lackOfCard = true
                }
            }
```

CardDeck struct를 class로 바꿔서 이를 상속해 Dealer 클래스를 만들었습니다. 이를 통해 CardDeck 클래스의 메소드를 이용해 딜러 클래스가 실제 딜러가 게임을 진행하 듯 구현할 수 있을 것 같았기 때문입니다.

jung-yun committed:
Edit String literal

jung-yun committed:
Fix pattern density error

jung-yun committed:
Edit README

jung-yun committed:
Implement CustomStringConvertible protocol

jung-yun committed:
Create Player structure

jung-yun committed:
Implement PokerGame class

jung-yun committed:
Edit PokerGame class and relevant interfaces

godrm commented on PokerGameApp/PokerGameApp/Assets.xcassets/bg_pattern.imageset/Contents.json:
```diff
@@ -10,6 +9,7 @@
     {
+      "filename" : "bg_pattern.png",
       "idiom" : "universal",
       "scale" : "3x"
```
배율을 바꿨는데 그 이유가 궁금합니다. 실제로 이미지 크기도 달라졌을까요?

jung-yun replied:
넵~! 1x로 지정했을땐 요구사항에 나와있는 것 보다 패턴이 크게 나타났는데 3x로 지정시 패턴이 좀 더 빽빽하게 나왔습니다. 아이폰 12에서 1pt는(3x3)픽셀이라 1x 이미지 같은 경우는 자동으로 가로세로 3배가 돼서 적용된다고 피어세션을 통해 알게됐습니다. (Thanks to Hong)

godrm commented on PokerGameApp/PokerGameApp/Card.swift:
```diff
@@ -36,7 +36,23 @@ class Card {
 }
 
+//MARK: -EXTENSIONS
+extension Card.Suit: CustomStringConvertible {
+    var description: String {
+        return "\(self.rawValue)"
+    }
+}
+extension Card.Value: CustomStringConvertible {
+    var description: String {
+        return "\(self.rawValue)"
+    }
+}
+extension Card: CustomStringConvertible {
+    var description: String {
+        return "\(self.cardDescription)"
+    }
+}
```
어떤 경우에 extension으로 타입을 확장해야 할까요? 이렇게 파일 하단에 모아놓는 게 더 좋은 표현일까요? 
cardDescription 속성은 불필요한 것 같습니다.  `"\(suit)\(value)"`를 전달해도 됩니다.

godrm commented on PokerGameApp/PokerGameApp/Player.swift:
```diff
@@ -0,0 +1,31 @@
+
+class Player {
+    var playerID: Int = 0
+    var playerCardDeck: Array<Card>?
```
내부에서 쓰는 이름에 모두 player를 붙일 껀가요? 그냥 Id, cardDeck 으로 써도 되지 않을까요? 
여기 파일도 Player고 타입도 Player인걸요

godrm commented on PokerGameApp/PokerGameApp/Player.swift:
```diff
@@ -0,0 +1,31 @@
+            playerCardDeck?.append(card)
+        } else {
+            playerCardDeck = [card]
+        }
```
초기화가 안된 경우 때문에 이렇게 구현한건가요?
init() 함수의 역할이 뭘까요? 바로 초기값을 지정하는 겁니다. 빈 배열을 init()에서 준비하면 되지 않을까요?

jung-yun replied:
확실히 빈배열을 init()에서 준비하니 좀 더 깔끔한 코드가 됐습니다. 감사합니다~!

godrm commented on PokerGameApp/PokerGameApp/PokerGame.swift:
```diff
@@ -0,0 +1,76 @@
+import Foundation
+
+class Dealer: CardDeck {
+    var dealerDeck: Array<Card>?
```
Dealer가 CardDeck을 상속받기 위해서 class로 바꾼 결정은 나쁜 결정이네요.
상속 관계보다 더 좋은 것은 포함하는 겁니다. OOP 수업할 때 상속은 같은 부류일 때만 상속을 한다고 했습니다. 
지금 관계는 무생물 카드덱을 딜러 사람이 상속한 경우네요. 
CardDeck 과 Dealer가 서로 어떤 포함 관계를 갖는 게 좋은지 고민해서 개선해보세요.

godrm commented on PokerGameApp/PokerGameApp/PokerGame.swift:
```diff
@@ -0,0 +1,76 @@
+    }
+}
+
+class pockerGame {
```
- 타입을 선언할 때는 대문자로 해주세요
- 서로 Dealer와 PokerGame은 서로 다른 타입이니까 파일을 분리해주세요.

jung-yun replied:
class를 소문자로 선언한 후 이니셜라이즈 단계에서 일어날 수 있는 상황에 대해 알게되었습니다 감사합니다~!

godrm commented on PokerGameApp/PokerGameApp/PokerGame.swift:
```diff
@@ -0,0 +1,76 @@
+    var cardDeck = CardDeck()
+    var dealer = Dealer()
+    var playersArray: Array<Player> = []
+    var lackOfCard: Bool = false
```
습관적으로 struct나 class 속성은 private var를 우선적으로 고려하세요.

godrm commented on PokerGameApp/PokerGameApp/PokerGame.swift:
```diff
@@ -0,0 +1,76 @@
+        print(cardDeck.count())
+    }
+    
+    func distribute(numberOfCards: Int) {
```
numberOfPlayers 나 numberOfCards 를 다룰 때도 Int 타입을 그대로 쓰기 보다는 참가자 인원수나 게임 종류를 의미하는 타입을 선언해보세요. 그 타입을 사용해보세요.

jung-yun replied:
Int 상수를 쓰지않고 타입선언을 고민하다보니 Enum타입을 쓰는 좀 더 깔끔하고 오류가 적은 코드를 쓰게된 것 같아요. 앞으로도 신경쓰면서 하겠습니다~!

godrm commented on PokerGameApp/PokerGameApp/PokerGame.swift:
```diff
@@ -0,0 +1,76 @@
+class pockerGame {
+    var cardDeck = CardDeck()
+    var dealer = Dealer()
+    var playersArray: Array<Player> = []
```
변수나 속성 이름에 Array, String 이런 타입을 붙이지 마세요. 대신 그 역할이 무엇인지 명시하는 게 좋습니다. 
덧붙여서 Array<Player>를 포함하는 Players 타입을 선언해보세요. 
CardDeck처럼 [Card]를 포함하는 타입으로 만들면 됩니다. 
그러면 여기서 playersArray 배열에 접근하는 코드들이 어떻게 이동하는 지 확인해보세요.

godrm commented:
PR 오픈해놓겠습니다. 피드백 남긴 내용들 확인해서 다시 PUSH해주세요.
모두 수정하고 나면 댓글을 남겨주세요.

jung-yun committed:
Refactor code1

jung-yun committed:
Re-construct Card class

jung-yun committed:
Reconstruct Dealer class

jung-yun committed:
Re-set access levels of member properties

jung-yun committed:
Refactor code2

jung-yun committed:
Add Players class

welcomePlayers method, by which number of players are set, moved to the new class
Changed the member property players into paticipants in PockerGame so to be distinguishable with that of Players class

jung-yun committed:
Fix typo

jung-yun commented:
@godrm, 수정완료했습니다~.

godrm commented on PokerGameApp/PokerGameApp/Card.swift:
```diff
@@ -34,25 +32,12 @@ class Card {
 extension Card: CustomStringConvertible {
     var description: String {
-        return "\(self.cardDescription)"
+        return "\(suit.rawValue)\(value.rawValue)"
```
Card: CustomStringConvertible  채택한 것처럼 Suit 와 Value도 CustomStringConvertible프로토콜을 채택해보세요.

godrm commented on PokerGameApp/PokerGameApp/Player.swift:
```diff
@@ -8,24 +8,21 @@
-        } else {
-            playerCardDeck = [card]
-        }
+        cardDeck.append(card)
```
Dealer와 Player가 구현이 동일한 메소드가 있는 것 같은데, 일반화 시켜보는 건 어떨까요?

godrm commented on PokerGameApp/PokerGameApp/Players.swift:
```diff
@@ -0,0 +1,18 @@
+    init(numberOfPlayers: Int) {
+        for i in 1...numberOfPlayers {
+            players.append(Player(id: i))
+        }
```
포인트는 단지 배열만 갖고 있고 public var로 players를 열어주는 게 아닙니다.
이러면 또 외부에서 Players.players 이렇게 접근해서 탐색하게 되니까요.
외부에서 players 배열에 접근하는 코드를 여기로 옮기셔야 합니다.

jung-yun committed:
Apply CustomStringConvertible to Card.Suit Card.Value respectively

jung-yun committed:
Implement Receiver Protocol

jung-yun committed:
Edit Players class and relevant codes

jung-yun commented:
Dealer 와 Players의 동일 메소드 일반화를 위해 Receiver 프로토콜을 작성했습니다. 처음에는 글로벌 함수와 Generic 타입을 이용해서 구현하는가 싶어 많이 헤맸던 것 같습니다.

그리고 Players에 내재화 할 수 있는 코드를 내재화했습니다. 해보니 코드가 정말 깔끔해졌습니다~!

godrm commented on PokerGameApp/PokerGameApp/CardDeck.swift:
```diff
@@ -9,7 +9,7 @@ import Foundation
 
 class CardDeck{
-    private var currentCardDeck: Array<Card> = []
+    public var currentCardDeck: Array<Card> = []
```
public으로 바꾼 부분은 아쉽습니다 :)

jung-yun replied:
이것저것 시도해보다 private를 public으로 바꿨나봐요... 다른 클래스에서 currentCardDeck에 접근하는 부분이 없는데 private로 못 바꾸고 있었네요ㅠㅠ 좀 더 꼼꼼하게 확인해야했는데 실수했습니다... 방금 public 수정 및 실행 확인 했습니다🙏

godrm commented:
근데 브랜치나 PR 제목은 step3 인데 step4를 한거 맞죠? ㅎㅎ

jung-yun commented:
step2,3 뒤죽박죽 진행했더니 브랜치 이름을 계속 잘못쓰고 있었네요 🤯
